drivers: spi: add Data Fusion Subsystem SPI driver

SiyuanCheng-CN opened the pull request:
Introduce DesignWare ARC Data Fusion IP Subsystem(DFSS) SPI driver for ARC boards, i.e. EMSDP, which uses DW SPI to control SPI-Flash and DFSS SPI to connect external devices.

tbursztyka reviewed (dismissed):
Can you explain why you copy-pasted spi_dw.* and renamed it to spi_dfss.*  to just modify it a tiny bit. Why not, instead, modifying/adapting the spi_dw.* version to make it work with your target ?

teburd reviewed (dismissed):
agree with @tbursztyka this doesn't make sense

ruuddw commented:
I assume it is because two instances of the driver are used on this board, one with a memory mapped interface, and one with a ARC AUX register interface (CONFIG_SPI_DW_ARC_AUX_REGS): compiling one version with and one without doesn't work with a single driver in a single binary.
But, indeed if somehow having two nearly identical duplicates can somehow be avoided that would be good.

evgeniy-paltsev commented:
@SiyuanCheng-CN could you please confirm that main difference to spi_dw driver is that we use AUX mapped registers here?

Do we have any other significant difference?

Commit:
drivers: spi: add Data Fusion Subsystem SPI driver

Introduce DesignWare ARC Data Fusion IP Subsystem(DFSS) SPI
driver for ARC boards, i.e. EMSDP, which uses DW SPI to controll
SPI-Flash and DFSS SPI to connect external devices. Both drivers
share most source code, but DFSS uses ARC auxiliary registers.
Move FIFO depth setting to device tree.

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

tbursztyka commented:
> But, indeed if somehow having two nearly identical duplicates can somehow be avoided that would be good.

We actually used to support that when Curie SoC was still a thing. (the x86 and arc cores had a different register mappings/sizes)

@SiyuanCheng-CN check the git history on this driver.

ruuddw commented:
Did the AUX (sensor subsystem) and memory mapped (x86) versions co-exist in a single binary for Curie?
I think ARC/sensor subsystem and quark had different Zephyr images.
But you're right, the AUX interface for closely coupled peripherals in ARC subsystems are similar, maybe there is something to learn from the history.

SiyuanCheng-CN commented:
@tbursztyka @teburd I've checked  [Curie SPI driver](https://github.com/tbursztyka/zephyr/blob/6f42f0a0c4eb8e066d5ed227a53f9895a5abe83a/drivers/spi/spi_dw.h#:~:text=%23ifdef%20CONFIG_SOC_QUARK_SE_C1000_SS,spi_dw_regs.h%22). looks like it only allows one version (Curie style or general style) to exist at the same time. Current status about EMSDP board is that it can use two versions of DW SPI ( [AUX interface and memory map)](https://github.com/tbursztyka/zephyr/blob/6f42f0a0c4eb8e066d5ed227a53f9895a5abe83a/drivers/spi/spi_dw.h#:~:text=%23ifdef%20CONFIG_SPI_DW_ARC_AUX_REGS,/*%20CONFIG_SPI_DW_ARC_AUX_REGS%20*/) simultaneously. 
So, at the very beginning, I thought this driver might become complex and become less efficient and turn to directly create a new driver. If not creating a new driver, you can check my current changes. It uses `aux_reg` to distinguish the version of driver currently using.

abrodkin reviewed (changes requested):
See comments below.

evgeniy-paltsev commented on drivers/spi/spi_dw.h:
```diff
@@ -19,13 +19,27 @@ extern "C" {
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pcfg;
+#endif
+	uint8_t aux_reg;
```
@abrodkin

A bit strange advise TBH. This is a flag, so let's just use `bool` type and allow compiler decide about the optimization.

abrodkin replied:
Agree, `bool` is even better!

tbursztyka replied:
just rearrange the attributes order so it's properly aligned (looks like 4bytes could be saved)

tbursztyka commented on drivers/spi/spi_dw.h:
```diff
@@ -35,52 +49,125 @@ struct spi_dw_data {
 	uint16_t _unused;
 };
 
+/* Register operation funcitons */
```
functions

tbursztyka commented on drivers/spi/spi_dw.h:
```diff
@@ -35,52 +49,125 @@ struct spi_dw_data {
-
 #define DEFINE_MM_REG_READ(__reg, __off, __sz)				\
-	static inline uint32_t read_##__reg(uint32_t addr)			\
+	static inline uint32_t read_##__reg(spi_dw_read_t read_func, uint32_t addr)	\
```
since the function is always info->read_func and since addr is always info->regs, let's switch to read_##__reg(struct spi_dw_config *info)
It will be cleaner in the driver side (it will keep the function calls short).

Apply the same on all macros below

tbursztyka reviewed (commented):
Much better. 
(btw _now_  you can add your copyright since the changes are quite substantial)

SiyuanCheng-CN commented on drivers/spi/spi_dw.c:
```diff
@@ -234,6 +238,12 @@ static int spi_dw_configure(const struct spi_dw_config *info,
 		return -EINVAL;
 	}
 
+	/* DFSS only*/
```
Added clock_on() and clock_off() functions, but seems like CLK_ENA register is not platform specific. Curie also uses this register with AUX space. I'm ok with both ways.

SiyuanCheng-CN replied:
Now changed to enable the CLK_ENA during kernel init.

abrodkin commented on boards/arc/emsdp/platform.c:
```diff
@@ -0,0 +1,22 @@
+
+static int emsdp_dfss_clock_init(void)
+{
+	sys_out32(1, DFSS_SPI0_BASE + DW_SPI_REG_CLK_ENA);
```
Even though from used defines one may conclude what's being done here, I'd suggest to add comments like "Enable clock for DFSS SPI0 controller" & "Enable clock for DFSS SPI1 controller".

abrodkin commented on boards/arc/emsdp/platform.c:
```diff
@@ -0,0 +1,22 @@
+
+#define DFSS_SPI0_BASE			0x80010000
+#define DFSS_SPI1_BASE			0x80010100
+#define DW_SPI_REG_CLK_ENA		(0x16)  /* DFSS only */
```
If it's not related to the DW SPI IP, maybe don't hint on that with the definition name `DW_SPI_xxx`?

abrodkin commented on drivers/spi/spi_dw.c:
```diff
@@ -4,6 +4,12 @@
  */
 
+/*
+ * Copyright (c) 2023 Synopsys
```
Why adding entire header here? Just add ` Copyright (c) 2023 Synopsys, Inc. All rights reserved.` in the 3-rd line above.

abrodkin commented on drivers/spi/spi_dw_regs.h:
```diff
@@ -6,6 +6,12 @@
  */
 
+/*
+ * Copyright (c) 2023 Synopsys
```
Ditto.

abrodkin reviewed (changes requested):
See some comments below.

abrodkin reviewed (changes requested):
See some comments below.

evgeniy-paltsev commented on boards/arc/emsdp/Kconfig.defconfig:
```diff
@@ -24,10 +24,7 @@ config SPI_DW
-
-config SPI_DW_ARC_AUX_REGS
-	default n
+	default 16
```
I'm wondering about why do we change fifo size for emsdp?

Was it incorrect? Or DW and DFSS peripherals have different FIFO size on emsdp?

SiyuanCheng-CN replied:
emsdp fifo depth is 32 for DW. 16 for DFSS.

abrodkin replied:
IMHO that needs to be moved to the .dts as the right place for HW description. That will code simpler and .dts more informative (more complete).

SiyuanCheng-CN commented:
Moved FIFO depth setting to device tree.

carlescufi commented:
@tbursztyka can you take another look? Thanks!

carlescufi commented:
> @tbursztyka can you take another look? Thanks!

Ah, I see you were only waiting for @abrodkin's comments, merging then.
